crates/re_chunk/src/chunk.rs

```diff
@@ -0,0 +1,478 @@
+use std::collections::BTreeMap;
+
+use arrow2::array::Array as ArrowArray;
+
+use re_log_types::{EntityPath, ResolvedTimeRange, RowId, TimeInt, TimePoint, Timeline};
+use re_types_core::{ComponentName, SerializationError};
+
+// ---
+
+/// Errors that can occur when creating/manipulating a [`Chunk`]s.
+#[derive(thiserror::Error, Debug)]
+pub enum ChunkError {
+    #[error("Detected malformed Chunk: {reason}")]
+    Malformed { reason: String },
+
+    #[error(transparent)]
+    Serialization(#[from] SerializationError),
+
+    #[error("Chunks cannot be empty")]
+    Empty,
+}
+
+pub type ChunkResult<T> = Result<T, ChunkError>;
+
+// ---
+
+/// Unique identifier for a [`Chunk`], using a [`re_tuid::Tuid`].
+pub type ChunkId = re_tuid::Tuid;
+
+/// Dense arrow-based storage of N rows of multi-component multi-temporal data for a specific entity.
+///
+/// This is our core datastructure for logging, storing, querying and transporting data around.
+///
+/// The chunk as a whole is always ascendingly sorted by [`RowId`] before it gets manipulated in any way.
+/// Its time columns might or might not be ascendingly sorted, depending on how the data was logged.
+///
+/// This is the in-memory representation of a chunk, optimized for efficient manipulation of the
+/// data within.
+#[derive(Debug, Clone)]
+pub struct Chunk {
+    pub(crate) id: ChunkId,
+    pub(crate) entity_path: EntityPath,
+
+    /// Is the chunk as a whole sorted by [`RowId`]?
+    pub(crate) is_sorted: bool,
+
+    /// The respective [`RowId`]s for each row of data.
+    pub(crate) row_ids: Vec<RowId>,
+
+    /// The time columns.
+    ///
+    /// Each column must be the same length as `row_ids`.
+    ///
+    /// Empty if this is a static chunk.
+    pub(crate) timelines: BTreeMap<Timeline, ChunkTimeline>,
+
+    /// A sparse `ListArray` for each component.
+    ///
+    /// Each `ListArray` must be the same length as `row_ids`.
+    ///
+    /// Sparse so that we can e.g. log a `Position` at one timestamp but not a `Color`.
+    pub(crate) components: BTreeMap<ComponentName, Box<dyn ArrowArray>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ChunkTimeline {
+    /// Every single timestamp for this timeline.
+    ///
+    /// * This might or might not be sorted, depending on how the data was logged.
+    /// * This is guaranteed to always be dense, because chunks are split anytime a timeline is
+    ///   added or removed.
+    /// * This can never contain `TimeInt::STATIC`, since static data doesn't even have timelines.
+    //
+    // TODO(cmc): maybe this would be better as raw i64s so getting time columns in and out of
+    // chunks is just a blind memcpy… it's probably not worth the hassle for now though.
+    // We'll see how things evolve as we start putting chunks in the backend.
+    pub(crate) times: Vec<TimeInt>,
```

Comment:
Depending on how the backend side goes, I might actually end up not deserializing these at all, if I can afford it. That would be sweet.